File: src/com/path/utils/PathUtils.java
package com.path.utils;

import lombok.Getter;
import lombok.Setter;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;

import java.io.File;
import java.io.IOException;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.nio.file.Paths;

@Getter
@Setter
public class PathUtils {
    private final String EMPTY_STRING;
    //Null String can be anything.
    private final String NULL_STRING;

    //Check to see if the generated path from methods exist in the computer or not, if not it will return empty String value.
    private boolean checkPath;
    //Enable to check if the input paths from methods are null or not, if it's true if will throw NullPointerException
    //and if it is false, it will return an empty String value for the path.
    private boolean throwNullPointerException;
    //If Dot(.) character detected in method input if true it will return the current path,application is executing
    //and if false it will return empty String value.
    private boolean disableDotCharacter;

    public PathUtils(boolean checkPath, boolean throwNullPointerException) {
        EMPTY_STRING = "";
        NULL_STRING = "null";

        this.checkPath = checkPath;
        this.throwNullPointerException = throwNullPointerException;
        this.disableDotCharacter = false;
    }

    /**
     * Returns the full path of a file or folder by its name.
     *
     * @param name The file or folder name.
     * @return The path of file or folder.
     */
    public String getPath(String name) {
        if (checkInput(name).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        if (checkPath) {
            return FilenameUtils.getPath(name);
        }

        if (disableDotCharacter && name.equals(".")) {
            return NULL_STRING;
        }

        return new File(name).getAbsolutePath();
    }

    /**
     * Returns only the name of a file or folder with its extension .
     *
     * @param path The file or folder path.
     * @return For the files it returns the full name with extension and for folders it returns just the name.
     */
    public String getName(String path) {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        return FilenameUtils.getName(path);
    }

    /**
     * Returns the file or folder name without extension.
     *
     * @param path The file or folder path.
     * @return For the files it returns only the file name without extension and for folders it returns just the name.
     */
    public String getBaseName(String path) {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        return FilenameUtils.getBaseName(path);
    }

    /**
     * Returns the extension of a file or folder.
     *
     * @param path The file or folder path.
     * @return For the files it returns only the extension of file without its name and for folders it returns just the name.
     */
    public String getExtension(String path) {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        return FilenameUtils.getExtension(path);
    }

    /**
     * Returns the parent folder name of specific path.
     *
     * @param path The file or folder path.
     * @return The parent name of file or folder.
     */
    public String getParentName(String path) {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        if (disableDotCharacter && path.equals(".")) {
            return NULL_STRING;
        }

        if (checkPath && !new File(path).exists()) {
            return NULL_STRING;
        }

        return Paths.get(path).getParent().getFileName().toString();
    }

    /**
     * Returns the parent folder path of specific path.
     *
     * @param path The file or folder path.
     * @return The parent full path of file or folder.
     */
    public String getParentPath(String path) throws IOException {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        if (disableDotCharacter && path.equals(".")) {
            return NULL_STRING;
        }

        if (checkPath && !new File(path).exists()) {
            return NULL_STRING;
        }

        return Paths.get(path).getParent().toRealPath(LinkOption.NOFOLLOW_LINKS).toString();
    }

    /**
     * Convert the specific path to normal path using separators and dots(.).
     *
     * @param path          The file or folder path.
     * @param unixSeparator The unix separator is used when you are using unix based os like mac or linux,for windows disable it.
     * @return The normalized path that can easily use.
     */
    public String normalize(String path, boolean unixSeparator) {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        String normalizedPath = FilenameUtils.normalize(path, unixSeparator);
        if (checkPath && !new File(normalizedPath).exists()) {
            return NULL_STRING;
        }

        return normalizedPath;
    }

    /**
     * Returns the file or folder new name without extension.
     *
     * @param path The path of file or folder.
     * @return For the files it will remove its extension and return the name of file
     * and for folders it doesn't do anything just returns the name of it.
     */
    public String removeExtension(String path) {
        if (checkInput(path).equals(NULL_STRING)) {
            return NULL_STRING;
        }

        return FilenameUtils.removeExtension(path);
    }

    /**
     * Check the methods input by specific arguments in the class, if NullPointerException is
     * set to true it will check for null inputs and throw exception if one of them is null otherwise it will return null string
     * or check to see if the generated path or name from methods actually exists in the computer or not, if true it will return it
     * and if false it will return null;
     *
     * @param input The input from methods arguments.
     * @return The filtered method inputs.
     */
    public String checkInput(String input) {
        if (input == null) {
            if (throwNullPointerException) {
                throw new NullPointerException("The input is null.");
            } else {
                return NULL_STRING;
            }
        }

        return input;
    }
}

File: src/main/kotlin/org/path/utils/PathUtils.kt
package org.path.utils

import org.apache.commons.io.FilenameUtils
import java.io.File
import java.nio.file.LinkOption
import java.nio.file.Paths

/**
 * @author Mahdi Lavasani (darkDev8)
 * This class helps you find file and folder paths easily and modify them how you want to be.
 * @param checkPath Enable it if you want to check the file or folder path before getting path
 * if you enable this and specific path can't be found, it will return empty result.
 * @param dotEnabled This future prevent interference of dot(.) character with function parameters(paths) because
 * dot(.) means same place the program executed and sometimes this make invalid results.
 * @param pathNotFound It will show this text if you can't find the specific file or folder path.
 */
class PathUtils(var checkPath: Boolean = false, var dotEnabled: Boolean = false, var pathNotFound: String = "null") {
    private val currentPath: String = System.getProperty("user.dir"); /* current path where this program executed*/
    private val emptyString: String = ""

    /**
     * Gets the path of file or folder.
     * @param name The name of file or folder you want to get its path.
     * @return If path found it will return it and if not it will return empty result.
     */
    fun getPath(name: String): String {
        val input = File(name)

        if (name == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> input.absolutePath
                false -> pathNotFound
            }
        }

        return input.absolutePath
    }

    /**
     * Get the file or folder name from path.
     * @param path The full path of file or folder you want to fetch its name.
     * @return The name of the file or folder.
     */
    fun getName(path: String): String {
        val input = File(path)

        if (path == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> FilenameUtils.getName(path)
                false -> pathNotFound
            }
        }

        return FilenameUtils.getName(path)
    }

    /**
     * Get the base name of file or folder.
     * The base name is only the name of file without extension and for folders, its name is base name.
     * @param path The full path of file or folder you want to fetch its base name.
     * @return The base name of file or folder.
     */
    fun getBaseName(path: String): String {
        val input = File(path)

        if (path == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> FilenameUtils.getBaseName(path)
                false -> pathNotFound
            }
        }

        return FilenameUtils.getBaseName(path)
    }

    /**
     * Get the extension of file or folder.
     * The extension of files, but folders don't have extension.
     * @param path The full path of file or folder you want to fetch its extension.
     * @return The extension of file or folder.
     */
    fun getExtension(path: String): String {
        val input = File(path)

        if (path == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> FilenameUtils.getExtension(path)
                false -> pathNotFound
            }
        }

        return FilenameUtils.getExtension(path)
    }

    /**
     * Get the folder name that this file or folder is in it.
     * @param path The full path of file or folder you want to fetch its parent name.
     * @return The parent folder name this file or folder is in it.
     */
    fun getParentName(path: String): String {
        val input = File(path)

        if (path == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> Paths.get(input.absolutePath).parent.fileName.toString()
                false -> pathNotFound
            }
        }

        return Paths.get(input.absolutePath).parent.fileName.toString();
    }

    /**
     * Get the folder full path that this file or folder is in it.
     * @param path The full path of file or folder you want to fetch its parent folder full path.
     * @return The parent folder full path this file or folder is in it.
     */

    fun getParentPath(path: String): String {
        val input = File(path)

        if (path == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> Paths.get(input.absolutePath).getParent().toRealPath(LinkOption.NOFOLLOW_LINKS).toString()
                false -> pathNotFound
            }
        }

        return Paths.get(input.absolutePath).getParent().toRealPath(LinkOption.NOFOLLOW_LINKS).toString()
    }

    /**
     * It will normalize the path and get you clean ready path to find files and folders.
     * It will process every folder in and out characters like (.) and (..)
     * @param path The full path of file or folder you want to fetch its parent folder full path.
     * @return After normalize it will return the new path.
     */
    fun normalize(path: String): String {
        val os = System.getProperty("os.name").lowercase()
        val input = if (os.contains("win")) File(FilenameUtils.normalize(File(path).absolutePath, false))
        else File(FilenameUtils.normalize(File(path).absolutePath, true))

        if (path == ".") {
            getDotResult()
        }

        if (checkPath) {
            return when (input.exists()) {
                true -> input.absolutePath
                false -> pathNotFound
            }
        }

        return input.absolutePath
    }

    /**
     * Check to see method input is dot(.) or not.
     * If dot future was enabled it will return current path the program executed
     * and if not enabled it will return not path found variable.
     * @return True path for dot character to prevent invalid path to create.
     */
    private fun getDotResult(): String {
        return when (dotEnabled) {
            true -> currentPath
            false -> pathNotFound
        }
    }
}